File: src/FE/src/app/modules/dashboard/services/dashboard.dto.ts
/**
 * Wire types (DTO) theo `doc/contracts/dashboard.md` (Status: IMPLEMENTED).
 * Bọc trong `ApiResponse<T>` (xem `core/http/api-response.model.ts`). Chỉ
 * dùng trong `dashboard.service.ts`.
 */

export interface DashboardKpiDto {
  OverallProgress: number | null;
  Delta: number | null;
  PreviousPeriodLabel: string | null;
  Up: number;
  Flat: number;
  Down: number;
  Done: number;
  TotalCriteria: number;
}

export interface DashboardGroupProgressDto {
  GroupId: string;
  GroupCode: string;
  GroupName: string;
  Progress: number | null;
}

export interface DashboardTrendPointDto {
  Label: string;
  Value: number | null;
}

export interface DashboardTableRowDto {
  CriteriaId: string;
  Code: string;
  Name: string;
  GroupCode: string;
  GroupName: string;
  MaxScore: number;
  PreviousValue: number | null;
  CurrentValue: number | null;
  Delta: number | null;
  Badge: string | null;
}

export interface DashboardResponseDto {
  Mode: string;
  PeriodLabel: string;
  Kpi: DashboardKpiDto;
  Groups: DashboardGroupProgressDto[];
  Trend: DashboardTrendPointDto[];
  Table: DashboardTableRowDto[];
}

export interface ReportResponseDto {
  Title: string;
  ContentHtml: string;
}

export interface PeriodOptionDto {
  Value: string;
  Date: string;
  OverallProgress: number | null;
}

export interface PeriodOptionsDto {
  Years: number[];
  WeeksInYear: PeriodOptionDto[];
  MonthsInYear: PeriodOptionDto[];
}

File: src/BE/Modules/DtiWeekly/PlatformManager.Modules.DtiWeekly.Application/Dashboard/DashboardDto.cs
namespace PlatformManager.Modules.DtiWeekly.Application.Dashboard;

public sealed record DashboardKpiDto(
    decimal? OverallProgress,
    decimal? Delta,
    int UpCount,
    int FlatCount,
    int DownCount,
    int DoneCount);

public sealed record GroupProgressDto(Guid GroupId, string GroupName, decimal? Progress);

public sealed record TrendPointDto(DateOnly WeekStart, decimal Progress);

/// <summary>Badge = "bdone"|"bstall"|"bwork"|"bnodata" — khớp 3 lớp CSS gốc của
/// dashboard.html + 1 trạng thái "chưa có dữ liệu" mới. Xem doc/ERD/ERD.md mục 4.</summary>
public sealed record CriteriaTableRowDto(
    Guid CriteriaId,
    string Code,
    string Name,
    Guid GroupId,
    string GroupName,
    decimal MaxScore,
    decimal? Progress,
    decimal? Delta,
    string Badge,
    string BadgeLabel);

public sealed class DashboardDto
{
    public bool HasData { get; init; }
    public required DashboardKpiDto Kpi { get; init; }
    public IReadOnlyList<GroupProgressDto> GroupProgress { get; init; } = [];
    public IReadOnlyList<TrendPointDto> Trend { get; init; } = [];
    public IReadOnlyList<CriteriaTableRowDto> Table { get; init; } = [];
}

/// <summary>1 lựa chọn kỳ trong dropdown — "value" truyền thẳng vào PeriodValue của
/// GET /api/criteria hoặc GET /api/dashboard ("2026-W33" cho tuần ISO, "2026-08" cho tháng).
/// Khớp IPeriodOptionDto ở src/FE/src/app/shared/models/period-options.model.ts (CONTRACT
/// DB-3/DM-8) — KHÔNG đổi lại thành mảng nguyên thuỷ.</summary>
public sealed record PeriodOptionDto(string Value, DateOnly Date, decimal? OverallProgress);

/// <summary>Danh sách kỳ có dữ liệu trong 1 năm — phục vụ dropdown chọn tuần/tháng ở FE
/// (dùng chung Dashboard + Danh mục DTI). GET /api/dashboard/periods.</summary>
public sealed class DashboardPeriodsDto
{
    /// <summary>Mọi năm có dữ liệu, LUÔN kèm năm hiện tại dù chưa có dữ liệu (CONTRACT DB-3).</summary>
    public IReadOnlyList<int> Years { get; init; } = [];
    public IReadOnlyList<PeriodOptionDto> WeeksInYear { get; init; } = [];
    public IReadOnlyList<PeriodOptionDto> MonthsInYear { get; init; } = [];
}
